components/script/dom/htmlelement.rs

```diff
@@ -897,6 +937,82 @@ impl HTMLElement {
             None => false,
         }
     }
+
+    /// <https://html.spec.whatwg.org/multipage/#rendered-text-fragment>
+    fn rendered_text_fragment(&self, input: DOMString) -> DomRoot<DocumentFragment> {
+        // Step 1: Let fragment be a new DocumentFragment whose node document is document.
+        let document = document_from_node(self);
+        let fragment = DocumentFragment::new(&document);
+
+        // Step 2: Let position be a position variable for input, initially pointing at the start
+        // of input.
+        let mut position = input.chars().peekable();
+
+        // Step : Let text be the empty string.
+        let mut text = String::new();
+
+        // Step 4
+        while let Some(ch) = position.next() {
+            match ch {
+                // While position is not past the end of input, and the code point at position is
+                // either U+000A LF or U+000D CR:
+                '\u{000A}' | '\u{000D}' => {
+                    if ch == '\u{000D}' && position.peek() == Some(&'\u{000A}') {
+                        // a \r\n pair should only generate one <br>,
+                        // so just skip the \r.
+                        position.next();
+                    }
+
+                    if !text.is_empty() {
+                        append_text_node_to_fragment(&document, &fragment, text);
+                        text = String::new();
+                    }
+
+                    let br = HTMLBRElement::new(local_name!("br"), None, &document, None);
+                    fragment.upcast::<Node>().AppendChild(br.upcast()).unwrap();
+                },
+                _ => {
+                    // Collect a sequence of code points that are not U+000A LF or U+000D CR from
+                    // input given position, and set text to the result.
+                    text.push(ch);
+                },
+            }
+        }
+
+        // If text is not the empty string, then append a new Text node whose data is text and node
+        // document is document to fragment.
+        if !text.is_empty() {
+            append_text_node_to_fragment(&document, &fragment, text);
+        }
+
+        fragment
+    }
+
+    /// <https://html.spec.whatwg.org/multipage/#merge-with-the-next-text-node>
+    fn merge_with_the_next_text_node(&self, node: DomRoot<Node>) {
```

Comment:
We could make this faillible, instead of calling it inside the `node.is::<Text>()` condition(or you `expect`  below and explain why it cannot fail).

Reply:
Not sure about making this fallible, from the spec it doesn't sound like an operation that I'd expect to fail. It seems more like a clean up step to make sure that we don't have multiple Text nodes next to each other, which is already the case if either `node` or `next` are something other than Text nodes, or there is no `next`.

Maybe we should also call `unwrap` on the Result of the call to `ReplaceData`, since from what I can tell it only fails when fed invalid UTF-16, and I'm not sure about the architecture of Servo but invalid UTF-16 probably shouldn't make its way into the DOM at all, so we should probably fail as soon as possible in that case, or what do you think?

Reply:
I've tried to rewrite/clean it up a bit, what do you think about this version?